File: Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        pairs = {')':'(', ']':'[', '}':'{'}
        for c in s:
            if c in pairs:
                if len(stack) > 0 and stack[-1] == pairs[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0

File: 2023/20. Valid Parentheses.cpp
class Solution {
public:
    unordered_map<char, char> pairs = {
        {']', '['},
        {')', '('},
        {'}', '{'}
    };
    bool isValid(string s) {
        stack<char> stack;
        for (char c : s) {
            if (pairs.find(c) != pairs.end()) {
                if (stack.empty() || stack.top() != pairs[c]) {
                    return false;
                }
                stack.pop();
            } else {
                stack.push(c);
            }
        }
        return stack.empty();
    }
};